### PR TITLE
HHH-13658 : make NO_PROXY unnecessary

### PR DESCRIPTION
**Pull request by sebersole**
- Better handle `FetchModeType#LAZY` for to-one associations based on whether bytecode-enhancement-as-proxy is enabled.  Minimize the cases a user is likely to need to use `@LazyToOne`
- See also EAP7-1402
- Currently has 5 failing tests, all related to assertions regarding handling of polymorphic to-one associations which is behavior these changes explicitly affect.  So most likely these are "expected"

### CODE REVIEW

**Comment by gbadner** [comment]
@sebersole, it looks like you are making this the default behavior for bytecode-proxies enabled in master branch. I think it's the right thing to do in master. I just want to confirm that's your intention.

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java [hunk 1]
```diff
@@ -25,11 +25,18 @@
...
+
 	/**
 	 * Should the given property be included in the owner's base fetch group?
 	 */
 	public static boolean includeInBaseFetchGroup(
 			Property bootMapping,
+			InheritanceChecker inheritanceChecker,
 			boolean isEnhanced,
 			boolean allowEnhancementAsProxy,
 			boolean collectionsInDefaultFetchGroupEnabled) {
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java:42]: I'm assuming that changing this SPI is OK in master.

Note to self: when this gets backported, the old version of the method should be deprecated.

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java [hunk 2]
```diff
@@ -51,15 +58,38 @@ public static boolean includeInBaseFetchGroup(
...
+						BytecodeLogger.LOGGER.debugf(
+								"`%s#%s` was mapped with LAZY with explicit NO_PROXY but the associated entity (`%s`) has subclasses",
+								bootMapping.getPersistentClass().getEntityName(),
+								bootMapping.getName(),
+								toOne.getReferencedEntityName()
+						);
 					}
-					// include it in the base fetch group so long as the config allows
-					// using the FK to create an "enhancement proxy"
+					// however, select the fk to create the proxy
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java:82]: In `DefaultLoadEventListener#proxyOrLoad` there is a check for the existence of a `ProxyFactory` [here](https://github.com/hibernate/hibernate-orm/blob/master/hibernate-core/src/main/java/org/hibernate/event/internal/DefaultLoadEventListener.java#L283). 

Do we need to check for a `ProxyFactory` here and factor that into the decision for the return value?

IIUC, this will apply regardless of how `allowEnhancementAsProxy` is set. I think that's fine for master.

When this gets backported, I don't think this should return `true` when `allowEnhancementAsProxy == false`. WDYT?

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java [hunk 3]
```diff
@@ -51,15 +58,38 @@ public static boolean includeInBaseFetchGroup(
...
+					return true;
+				}
+
+				if ( toOne.isUnwrapProxy() && ! toOne.isUnwrapProxyImplicit() ) {
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java:86]: I'm unsure about this `if` statement. I'm going to do more checking on this Monday.

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/LazyAttributesMetadata.java [hunk 4]
```diff
@@ -35,7 +37,8 @@ public static LazyAttributesMetadata from(
 			PersistentClass mappedEntity,
 			boolean isEnhanced,
 			boolean allowEnhancementAsProxy,
-			boolean collectionsInDefaultFetchGroupEnabled) {
+			boolean collectionsInDefaultFetchGroupEnabled,
+			PersisterCreationContext creationContext) {
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/LazyAttributesMetadata.java:41]: I'm assuming that changing this SPI is OK in master.

Note to self: when this gets backported, the old version of the method should be deprecated.

File: hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java [hunk 5]
```diff
@@ -3163,7 +3163,9 @@ else if ( oneToOne != null ) {
 		}
 		else {
 			toOne.setLazy( fetchType == FetchType.LAZY );
-			toOne.setUnwrapProxy( false );
+
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java:3166]: Should the following be set to `true` when ` fetchType != FetchType.LAZY`

Will this change behavior for non-lazy attributes that end up getting loaded as a proxy from the `PersistenceContext`?

I'd like to look more deeply at this on Monday.

File: hibernate-core/src/main/java/org/hibernate/tuple/PropertyFactory.java [hunk 6]
```diff
@@ -148,14 +150,16 @@ public static VersionProperty buildVersionProperty(
...
 	 * @return The appropriate NonIdentifierProperty definition.
 	 */
 	public static NonIdentifierAttribute buildEntityBasedAttribute(
 			EntityPersister persister,
 			SessionFactoryImplementor sessionFactory,
 			int attributeNumber,
 			Property property,
-			boolean lazyAvailable) {
+			boolean lazyAvailable,
+			PersisterCreationContext creationContext) {
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/tuple/PropertyFactory.java:162]: I'm assuming that changing this SPI is OK in master.

Note to self: when this gets backported, the old version of the method should be deprecated.

File: hibernate-core/src/main/java/org/hibernate/tuple/entity/EntityMetamodel.java [hunk 7]
```diff
@@ -131,8 +133,8 @@
 	public EntityMetamodel(
 			PersistentClass persistentClass,
 			EntityPersister persister,
-			SessionFactoryImplementor sessionFactory) {
-		this.sessionFactory = sessionFactory;
+			PersisterCreationContext creationContext) {
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/tuple/entity/EntityMetamodel.java:136]: I'm assuming that changing this SPI is OK in master.

Note to self: when this gets backported, the old version of the method should be deprecated.

File: hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java (hunk 5 above)
Comment (pblanchardie) [reply hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java:3166]: I think yes. I tried, and it fixed the bug ;)

**Review by beikov — COMMENTED** [review]
Since all uses of `includeInBaseFetchGroup` negate the result to determine if the property is lazy, it's a bit hard to follow the logic for me. If this needs changes anyway, could we maybe rename this to `isPropertyLazy` and remove the negations?

Also, do I understand this correct, that lazy in this context means that the entity field/property does not receive a proxy? I conclude that from the fact that `to-one` associations to non-primary keys are always treated as "lazy" for which we can never have proxies, because proxies can only use the primary key. If this understanding is correct, I think the term _lazy_ is still a bit confusing when considering that associations can be lazy as well even if the fields/properties receive a proxy. How about using an enum then and naming the method `getPropertyInitializationMode` with options `NULL`, `PROXY` and `VALUE`. This would be a great help for my understanding of what is going on and as far as I understood (correct me if I am wrong), this is essentially what we are trying to model here.

**Comment by sebersole** [comment]
> @sebersole, it looks like you are making this the default behavior for bytecode-proxies enabled in master branch. I think it's the right thing to do in master. I just want to confirm that's your intention.

That is the way it is defined now in this work.  We can talk about whether that makes sense for 5.3 (anecdotally, I think it does).
The main problem with not (and this is the main difficulty with this feature at all in 5.x) is that its made in multiple places that sometimes do not have all the needed info.

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java (hunk 1 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java:42]: I really cannot imagine user call this.  But sure, whatever you think wrt 5.3

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java (hunk 2 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java:82]: That's a good question.  The thought process here was that the check for ProxyFactory was really just needed for entities with subclasses.  That is maybe not exactly correct yet.  This is obviously the main point of change for this improvement and its tricky to get just right - it comes back to what I mentioned earlier about timing and access to info.  This block is exactly why this is still a draft :)

Keep in mind also that this simply defines what columns should be selected as part of the "base fetch group".  I'm not sure I agree with your statement about `allowEnhancementAsProxy == false`.  Still need to play with it some more

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java (hunk 3 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java:86]: This check is actually the exact improvement.  Previously, if `@LazyToOne` was not specified (implicit) then unwrap-proxy would be set to true.  However, if `@LazyToOne` is implicit we want to treat that differently.

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/LazyAttributesMetadata.java (hunk 4 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/LazyAttributesMetadata.java:41]: Again, I cannot see this being a method users will call.  This is specifically creating these objects

File: hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java (hunk 5 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java:3166]: I think that is probably true

File: hibernate-core/src/main/java/org/hibernate/tuple/PropertyFactory.java (hunk 6 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/tuple/PropertyFactory.java:162]: Same comment as previous

File: hibernate-core/src/main/java/org/hibernate/tuple/entity/EntityMetamodel.java (hunk 7 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/tuple/entity/EntityMetamodel.java:136]: Again, same as previous.  For this though *maybe* other Hibernate projects use it, but I doubt it

**Comment by sebersole** [comment]
> Since all uses of `includeInBaseFetchGroup` negate the result to determine if the property is lazy, it's a bit hard to follow the logic for me. If this needs changes anyway, could we maybe rename this to `isPropertyLazy` and remove the negations?

I prefer it to remain this way.  Your phrasing is just as confusing imo.  Lazy in what sense, which is what this comes down to.  The problem has always been that we have 2 distinct concepts that we call "lazy" so that phrasing gets confusing.


> Also, do I understand this correct, that lazy in this context means that the entity field/property does not receive a proxy?

Not really, it means whether that property is considered part of the entity's base fetch group.  When we first initialize as entity, what state should we select from the database?  That's what this method determines.  That is not the same as whether there is a proxy or not.

I also thought of a different return value here.  I thought of an enum also (great minds...).  The problem is that of compatibility.  I think we should absolutely go that route with 6.0.  But for 5.x, to a large degree, it comes down to how much change in these signatures we want to introduce.

**Comment by sebersole** [comment]
> Not really, it means whether that property is considered part of the entity's base fetch group. When we first initialize as entity, what state should we select from the database?

To clarify a bit.. there are 3 conditions that need to be considered:

1. No enhancement
2. Enhanced but enhanced proxies disallowed
3. Enhanced and enhanced proxies allowed

The outcome here is generally different for all 3 cases.

**Comment by gbadner** [comment]
@sebersole, thanks for the feedback. I'll dig into this more today.

**Comment by gbadner** [comment]
I'm seeing 20 failures locally:

org.hibernate.jpa.test.callbacks.PrivateConstructorEnhancerTest > testGetReferenceEntity FAILED
    org.hibernate.testing.junit4.CallbackException at TestClassMetadata.java:208
        Caused by: javax.persistence.PersistenceException at PrivateConstructorEnhancerTest.java:46
            Caused by: org.hibernate.PersistentObjectException at PrivateConstructorEnhancerTest.java:46

org.hibernate.jpa.test.callbacks.PrivateConstructorEnhancerTest > testListEntity FAILED
    org.hibernate.testing.junit4.CallbackException at TestClassMetadata.java:208
        Caused by: javax.persistence.PersistenceException at PrivateConstructorEnhancerTest.java:46
            Caused by: org.hibernate.PersistentObjectException at PrivateConstructorEnhancerTest.java:46

org.hibernate.jpa.test.callbacks.PrivateConstructorEnhancerTest > testListJoinFetchEntity FAILED
    org.hibernate.testing.junit4.CallbackException at TestClassMetadata.java:208
        Caused by: javax.persistence.PersistenceException at PrivateConstructorEnhancerTest.java:46
            Caused by: org.hibernate.PersistentObjectException at PrivateConstructorEnhancerTest.java:46

org.hibernate.jpa.test.callbacks.PrivateConstructorEnhancerTest > testLoadProxyAssociation FAILED
    org.hibernate.testing.junit4.CallbackException at TestClassMetadata.java:208
        Caused by: javax.persistence.PersistenceException at PrivateConstructorEnhancerTest.java:46
            Caused by: org.hibernate.PersistentObjectException at PrivateConstructorEnhancerTest.java:46

org.hibernate.orm.test.mapping.lazytoone.InstrumentedJoinedLazyToOneTest > testEnhancedButProxyNotAllowed FAILED
    java.lang.AssertionError at InstrumentedJoinedLazyToOneTest.java:96

org.hibernate.orm.test.mapping.lazytoone.InstrumentedLazyToOneTest > testEnhancedButProxyNotAllowed FAILED
    java.lang.AssertionError at InstrumentedLazyToOneTest.java:88

org.hibernate.test.bytecode.enhancement.lazy.proxy.LazyGroupWithInheritanceTest > loadEntityWithAssociationToAbstract FAILED
    java.lang.AssertionError at LazyGroupWithInheritanceTest.java:57

org.hibernate.test.bytecode.enhancement.lazy.proxy.LazyGroupWithInheritanceTest > queryEntityWithAssociationToAbstract FAILED
    java.lang.AssertionError at LazyGroupWithInheritanceTest.java:88

org.hibernate.test.bytecode.enhancement.lazy.LazyLoadingIntegrationTest > test FAILED
    java.lang.NullPointerException at LazyLoadingIntegrationTest.java:82

org.hibernate.test.bytecode.enhancement.lazy.LazyLoadingTest > test FAILED
    java.lang.AssertionError at LazyLoadingTest.java:87

org.hibernate.test.bytecode.enhancement.lazy.LazyCollectionDetachTest > testDetachProxy FAILED
    javax.persistence.PersistenceException at LazyCollectionDetachTest.java:85
        Caused by: org.hibernate.HibernateException at LazyCollectionDetachTest.java:85
            Caused by: java.lang.NoSuchMethodException at LazyCollectionDetachTest.java:85

org.hibernate.test.bytecode.enhancement.lazy.QueryScrollingWithInheritanceTest > testScrollableWithSession FAILED
    java.lang.AssertionError at QueryScrollingWithInheritanceTest.java:160

org.hibernate.test.bytecode.enhancement.lazy.QueryScrollingWithInheritanceTest > testScrollableWithStatelessSession FAILED
    java.lang.AssertionError at QueryScrollingWithInheritanceTest.java:106

org.hibernate.test.bytecode.enhancement.lazy.NaturalIdInUninitializedAssociationTest > testImmutableNaturalId FAILED
    java.lang.AssertionError at NaturalIdInUninitializedAssociationTest.java:57

org.hibernate.test.bytecode.enhancement.lazy.NaturalIdInUninitializedAssociationTest > testMutableNaturalId FAILED
    java.lang.AssertionError at NaturalIdInUninitializedAssociationTest.java:74

org.hibernate.test.bytecode.enhancement.lazy.LazyCollectionLoadingTest > testNoTransaction FAILED
    org.hibernate.HibernateException at LazyCollectionLoadingTest.java:111
        Caused by: java.lang.NoSuchMethodException at LazyCollectionLoadingTest.java:111

org.hibernate.test.bytecode.enhancement.lazy.LazyCollectionLoadingTest > testTransaction FAILED
    org.hibernate.HibernateException at LazyCollectionLoadingTest.java:90
        Caused by: java.lang.NoSuchMethodException at LazyCollectionLoadingTest.java:90

org.hibernate.test.bytecode.enhancement.lazy.group.LazyGroupTest > testAccess FAILED
    java.lang.AssertionError at LazyGroupTest.java:104

org.hibernate.test.bytecode.enhancement.lazy.group.LazyGroupTest > testUpdate FAILED
    java.lang.AssertionError at LazyGroupTest.java:145

org.hibernate.test.bytecode.enhancement.ondemandload.OnDemandLoadTest > testClosedSession FAILED
    java.lang.AssertionError at OnDemandLoadTest.java:99

10304 tests completed, 20 failed, 1330 skipped

**Comment by sebersole** [comment]
> I'm seeing 20 failures locally:

You could have just seen those in the CI runs ;)

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java (hunk 2 above)
Comment (gbadner) [reply hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java:82]: OK, I see that the existence of a {{ProxyFactory}} isn't known at the time this is called.

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java [hunk 8]
```diff
@@ -39,7 +46,7 @@ public static boolean includeInBaseFetchGroup(
 			if ( value instanceof ToOne ) {
 				if ( ( (ToOne) value ).isUnwrapProxy() ) {
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java:47]: Due to the change below to `AnnotationBinder#defineFetchingStrategy`, if `@LazyToOne` is not mapped, then the following will be set:
```
toOne.setUnwrapProxy( true );
toOne.setUnwrapProxyImplicit( true );
```
and the message will end up getting DEBUG logged.

I think it should only get logged if NO_PROXY is explicitly mapped:

```
final ToOne toOne = (ToOne) value;
if ( toOne.isUnwrapProxy() && !toOne.isUnwrapProxyImplicit() ) {
```

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java [hunk 9]
```diff
@@ -51,15 +58,38 @@ public static boolean includeInBaseFetchGroup(
...
+				if ( inheritanceChecker.hasSubclasses( toOne.getReferencedEntityName() ) ) {
+					// the associated type has subclasses - we cannot use the enhanced proxy and will generate a HibernateProxy
+					if ( ! toOne.isUnwrapProxyImplicit() ) {
+						// NO_PROXY was explicitly requested
+						BytecodeLogger.LOGGER.debugf(
+								"`%s#%s` was mapped with LAZY with explicit NO_PROXY but the associated entity (`%s`) has subclasses",
+								bootMapping.getPersistentClass().getEntityName(),
+								bootMapping.getName(),
+								toOne.getReferencedEntityName()
+						);
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java:80]: I know it's a little late to ask this, but should this be a warning, or at least something a little stronger than just DEBUG logged, since `@LazyToOne(NO_PROXY)` is not compatible with an association with an entity class that has subclasses?

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java (hunk 3 above)
Comment (gbadner) [reply hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java:86]: OK, I understand this case (explicitly has NO_PROXY) now, so this looks OK.

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java [hunk 10]
```diff
@@ -51,15 +58,38 @@ public static boolean includeInBaseFetchGroup(
...
+				if ( toOne.isUnwrapProxy() && ! toOne.isUnwrapProxyImplicit() ) {
 					return allowEnhancementAsProxy;
 				}
 
+//				if ( toOne.isUnwrapProxy() || toOne.isUnwrapProxyImplicit() ) {
+//					return allowEnhancementAsProxy;
+//				}
 			}
 
 			return true;
```
Comment (gbadner) [inline hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java:95]: A lazy to-one association that does not explicitly have `@LazyToOne(NO_PROXY)`, and does not fall into the other special cases will end up returning `true`.

That makes sense for `allowEnhancementAsProxy == true`.

Is this correct if `allowEnhancementAsProxy == false`? Does that mean that the associated entity will be loaded as an uninitialized `HibernateProxy`?

File: hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java (hunk 5 above)
Comment (gbadner) [reply hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java:3166]: This is fine.

**Comment by gbadner** [comment]
@sebersole, I checked all the places where `ToOne#setUnwrapProxy` gets called, to make sure that the corresponding `ToOne#isUnwrapProxyImplicit` gets set properly.

I see that the calls from `ModelBinder` don't set `ToOne#isUnwrapProxyImplicit`. Since `ToOne#isUnwrapProxyImplicit` defaults to `false`, that DEBUG message in `EnhancementHelper#includeInBaseFetchGroup` will log the message for HBM mappings.

`OneToOneSecondPass#doSecondPass` also calls  `ToOne#setUnwrapProxy` without calling `ToOne#isUnwrapProxyImplicit` [here](https://github.com/hibernate/hibernate-orm/blob/master/hibernate-core/src/main/java/org/hibernate/cfg/OneToOneSecondPass.java#L199). I think this one is a bug that needs to be fixed.

File: hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java (hunk 5 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java:3166]: What I found is that knowing when the LazyToOne was implicit is just as important.  To that end, this is now as follows in my newest work, which I'll be pushing as a new PR soon -

```
toOne.setUnwrapProxy( fetchType != FetchType.LAZY );
toOne.setUnwrapProxyImplicit( true );
```

this new "implicitness" is checked later.  Ideally unwrap-proxy would have been a Boolean value allowing null rather than a boolean forcing a true/false here.  But I felt adding a new method was less disruptive since that mapping model is considered an API

**Comment by sebersole** [comment]
Closing in preference of https://github.com/hibernate/hibernate-orm/pull/3739
